README
NAME
    Net::Netfilter::NetFlow - App to generate Cisco NetFlow from Linux
    netfilter

PURPOSE
    If you run a Linux based firewall, the "nfflowd" application in this
    distribution will generate a stream of Cisco NetFlow logging data for
    all TCP, UDP and ICMP connections passing through.

    What makes this different from other solutions is that it uses Linux
    netfilter's own connection tracking data, rather than observing a
    traffic capture. This is more efficient for your CPU, and allows
    tracking of NAT which would otherwise not be possible.

    No reconfiguration of the firewall ruleset or network is required.

SYNOPSIS
    Make sure you have the "conntrack" and "flow-tools" packages installed
    on your Linux firewall system.

     write a small configuration describing your NetFlow collector:
     ~# cat > /etc/nfflowd.conf
     <flow_send>
         # source interface on this server
         args 172.16.1.1
         # netflow collector address
         args 192.168.0.1
         # netflow collector port number
         args 65001
     </flow_send>
     ^D
     
 to start the logger as a daemon:
     ~# nfflowd start
     
 to specify a custom configuration other than /etc/nfflowd.conf:
     ~# nfflowd -c /other/config/file.ini

LIMITATIONS
    *   It's assumed that Source IP NAT is taking place. If not, you may
        receive additional unecessary log lines, sorry. Work is planned to
        change the conntrack parser to be configurable for different local
        modes of operation.

    *   Currently only TCP, UDP and ICMP traffic is logged. Again this will
        likely be improved as part of the planned changes to the conntrack
        output parser.

    *   The tool supports only IPv4 tracking. Is there a "conntrack" for
        IPv6?

DESCRIPTION
    The application "nfflowd" included in this distribution is designed to
    run as a daemon (background) process on a Linux based firewall running
    IPtables.

    Its purpose is to capture data from the Linux netfilter subsystem about
    each tracked data connection passing through the firewall, and log this
    in Cisco NetFlow format to a remote NetFlow collector server. We
    recommend the Argus 3 software for that collector
    (<http://www.qosient.com/argus/>). With the default configuration you
    will also receive a local copy of the same data via "syslog", under the
    "local6" facility.

    "nfflowd" was written for NAT environments where you have a requirement
    to trace the IP address on the "private" side which made a particular
    connection or flow to a port/destination on the "public" (Internet)
    side. The NetFlow streams logged will allow you to trace back to that
    private IP.

    The daemon requires no reconfiguration of your firewall ruleset or
    network, but you may need to install one or two additional helper
    programs. These are commonly available in most Linux distributions - see
    "DEPENDENCIES", below.

    "nfflowd" is also efficient and accurate, as it uses Linux netfilter's
    own connection logging and tracking data. This is preferable to
    observing a promiscuous traffic capture, which may be CPU intensive and
    certainly cannot track NAT connections.

  What's special about NAT?
    An issue that's specific to NAT (rather than general Netflow from a
    Linux router) is that you really need a good and accurate record of the
    IP and port number translation. So there are *two* traffic flows, one
    before and one after the NAT. Both need logging, together with a way to
    correlate them.

    Traditionally such things are attempted using two instances of
    "softflowd" or "argus", one each side of the NAT. With accurate
    timestamping the two flows (pre and post-NAT) can be correlated and then
    the IP and port translation deduced.

    However this doesn't work for very long running flows, or where there
    are numerous flows to a destination, some of which are legitimate and
    some not. There exist cases where it is impossible to be sure that you
    are correlating the correct two flow records. Seeing as netfilter is
    already accurately tracking every connection, with NAT information, we
    can make use of that data.

USAGE
    The "nfflowd" application included in this distribution is designed to
    run as a daemon (background) process. It also supports running in the
    foreground. You will need to provide a small amount of configuration,
    and otherwise there are sane defaults for all options.

    Here is the help text for the current version of "nfflowd":

     Usage: nfflowd [ -c file ] [ -f ] { start | stop | restart | install | check | help | version } 
      -c file     Specify configuration file (instead of /etc/nfflowd.conf)
      -f          Run in the foreground (don't detach)
      start       Starts a new nfflowd if there isn't one running already
      stop        Stops a running nfflowd
      restart     Stops a running nfflowd if one is running.  Starts a new one.
      install     Setup nfflowd to run automatically after reboot
      check       Check the configuration file and report the daemon state
      help        Display this usage info
      version     Display the version of nfflowd

    Although the "reload" option is also available, it currently has no
    effect. Please note that the "install" option has not been tested by the
    author.

  Flow Output
    For each flow observed by "conntrack", three flow records are output by
    this tool. As explained below, three records are required to correlate
    the NAT operation which took place. The assumption of source IP NAT will
    be relaxed in a future release of this software.

    Here's an example of the three records generated by an HTTP requesst,
    which we'll walk through:

     1248952480,737235000,279737,172.16.1.1,720,38052,136315,279737,10.0.195.106,74.125.6.225,0.0.0.0,0,0,62380,80,6,0,0
     1248952480,737235000,279737,172.16.1.1,720,38052,136315,279737,192.76.7.154,74.125.6.225,0.0.0.0,0,0,62380,80,6,0,0
     1248952480,737235000,279737,172.16.1.1,1432,2027040,136315,279737,74.125.6.225,192.76.7.154,0.0.0.0,0,0,80,62380,6,0,0

    The format of these lines is "flow-tools" CSV for NetFlow version 5. Of
    course your NetFlow collector receives a proper UDP NetFlow payload, but
    here we are looking at the default Syslog output. The fields are listed
    in order in the table below:

       1248952480 UNIX_SECS  - Current count of seconds since 0000 UTC 1970
        737235000 UNIX_NSECS - Residual nanoseconds since 0000 UTC 1970
           279737 SYSUPTIME  - Current time in milliseconds since the export device booted
       172.16.1.1 EXADDR     - Export device IP address
              720 DPKTS      - Packets in the flow
            38052 DOCTETS    - Total number of Layer 3 bytes in the packets of the flow
           136315 FIRST      - SysUptime at start of flow
           279737 LAST       - SysUptime at the time the last packet of the flow was received
     10.0.195.106 SRCADDR    - Source IP address
     74.125.6.225 DSTADDR    - Destination IP address
          0.0.0.0 NEXTHOP    - IP address of next hop router
                0 INPUT      - SNMP index of input interface
                0 OUTPUT     - SNMP index of output interface
            62380 SRCPORT    - TCP/UDP source port number or equivalent
               80 DSTPORT    - TCP/UDP destination port number or equivalent
                6 PROT       - IP protocol type (for example, TCP = 6; UDP = 17)
                0 TOS        - IP type of service (ToS)
                0 TCP_FLAGS  - Cumulative OR of TCP flags

    The first line of the flow triplet describes the connection (IP/port)
    source and desination from the perspective of the originating host on
    the "private" side (or behind) your NAT.

    The second line describes how the source NAT was applied by netfilter.
    In this case the original host's source IP was translated from
    10.0.195.106 to 192.76.7.154, and the source port stayed the same
    (62380).

    The final line describes the return flow from that connection, so you
    can see how much traffic passed in each direction (NetFlow flows are
    unidirectional, so you do always get two flows for each connection
    anyway). We only log the public (post-NAT) IPs and port numbers, as
    that's enough data to complete the correlation.

    So in this example, host 10.0.195.106 make a TCP connection on port 80
    to 74.125.6.225, sending 38052 bytes and receiving 2027040 bytes back.

CONFIGURATION
    At minimum you need to provide a small configuration to let "nfflowd"
    know where your NetFlow collector server is. This by default can be
    located at "/etc/nfflowd.conf" although you can specify an alternate
    location using the "-c" command line option. The format of the file can
    be anything supported by Config::Any (i.e. YAML, JSON, etc), although we
    recommend using Config::General format, as in the examples below.

    The required configuration is the source IP on your firewall from which
    NetFlow packets are sent, and the IP/port numbers of the NetFlow
    collector server:

     <flow_send>
         # source interface on this server
         args 172.16.1.1
         # netflow collector address
         args 192.168.0.1
         # netflow collector port number
         args 65001
     </flow_send>

    These three options must appear *in the exact order given above*.

    The other thing you might want to configure is whether and how a local
    copy of the NetFlow data is stored on the Linux firewall itself, via
    Syslog. The application uses Log::Dispatch and by default will log to
    Syslog facility "local6" with priority <info>.

    Below is the default configuration, which you can override in your local
    "nfflowd" configuration file, but be warned! You *must* retain the
    "screen" dispatcher and configuration section, otherwise the application
    will stop working. You have the option to remove, change, or add to the
    "syslog" dispatcher only:

     <ptee>
         <conf>
             dispatchers screen
             dispatchers syslog
             <screen>
                 stderr    0
                 class     Log::Dispatch::Screen
                 min_level debug
             </screen>
             <syslog>
                 class     Log::Dispatch::Syslog
                 ident     nfflowd
                 min_level debug
                 facility  local6
             </syslog>
         </conf>
     </ptee>

    "nfflowd" records the start and end of each connection reported by
    netfilter. To preserve memory, the application clears up its records of
    "unfinished" connections which are older than a set TTL. This defaults
    to one week, but can be changed by setting a seconds value using the
    following configuration:

     <ct2ft>
         ttl  604800
     </ct2ft>

    In addition, you can control how each of the helper applications are
    loaded and configured. You would generally not need to do this, but here
    are the config options anyway:

     <conntrack>
         progname    conntrack
         init_format -L
         format      -E -e NEW,DESTROY -o timestamp -n
     </conntrack>
     <flow_import>
         progname flow-import
         format   -z0 -f2 -V5 -m0xFF31EF
     </flow_import>
     <flow_send>
         progname flow-send
         format   -V5 -s %s/%s/%s
     </flow_send>

PAR Support
    Simple tests have shown it should be possible to package up this
    application with PAR, and ship to servers which do not have Perl and/or
    the Perl module dependencies installed.

DEPENDENCIES
    Other than the standard contents of a Perl 5.8 distribution, you will
    need:

    *   conntrack at <http://conntrack-tools.netfilter.org/>

    *   flow-tools at <http://www.splintered.net/sw/flow-tools/>

    *   Log::Dispatch::Configurator::Any version 1.0005 or later

    *   Daemon::Generic

    *   File::Slurp version 9999.06 or later

    *   IPC::Run

    *   Config::General

    *   Config::Any version 0.15 or later

SEE ALSO
    "pfflowd" at <http://www.mindrot.org/projects/pfflowd/>
        This is the BSD pf equivalent of this tool.

    "softflowd" at <http://www.mindrot.org/projects/softflowd/>
        From the "pfflowd" stable, this tool requires a promiscuous traffic
        capture to generate data. It is therefore not fully stateful and
        cannot track NAT.

    http://www.cisco.com/en/US/tech/tk812/tsd_technology_support_protocol_ho
    me.html
        Cisco homepage for their NetFlow technology.

THANKS
    David Ford of the University of Oxford's OxCERT team provided valuable
    input and testing infrastructure, many thanks.

AUTHOR
    Oliver Gorwits "<oliver.gorwits@oucs.ox.ac.uk>"

COPYRIGHT & LICENSE
    Copyright (c) The University of Oxford 2009.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
{{$NEXT}}

1.113260  2011-11-22 22:22:23 Europe/London

  * Port to Dist::Zilla

libnet-netfilter-netflow-perl (0.07) UNRELEASED; urgency=low

  * Add more POD about the NetFlow output format 

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Thu, 30 Jul 2009 13:36:07 +0100

libnet-netfilter-netflow-perl (0.06) UNRELEASED; urgency=low

  * Add POD 

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Wed, 29 Jul 2009 14:43:37 +0100

libnet-netfilter-netflow-perl (0.05) UNRELEASED; urgency=low

  * Fixes to Daemon::Generic interfering with STDOUT/STDERR and upsetting IPC::Run

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Tue, 28 Jul 2009 20:49:27 +0100

libnet-netfilter-netflow-perl (0.04) UNRELEASED; urgency=low

  * Fix bugs in use of Config::General with File::Slurp 

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Tue, 28 Jul 2009 19:06:20 +0100

libnet-netfilter-netflow-perl (0.03) UNRELEASED; urgency=low

  * Change from File::ShareDir to useing <DATA> beacuse of PAR troubles

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Tue, 28 Jul 2009 18:44:14 +0100

libnet-netfilter-netflow-perl (0.02) UNRELEASED; urgency=low

  * Switch to Daemon::Generic to manage background running

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Tue, 28 Jul 2009 17:11:21 +0100

libnet-netfilter-netflow-perl (0.01) UNRELEASED; urgency=low

  * Initial release.

 -- Oliver Gorwits <oliver.gorwits@oucs.ox.ac.uk>  Tue, 28 Jul 2009 12:55:21 +0100


lib/Net/Netfilter/NetFlow/Utils.pm
package Net::Netfilter::NetFlow::Utils;

use strict;
use warnings FATAL => 'all';

use base 'Exporter';
our @EXPORT = qw(
    load_config
    format_args
    can_run
    merge_hashes
);

use Config::Any 0.15;

# use Config::Any to load a configuration file
sub load_config {
    my $name = shift;
    die "No filename specified to load_config!\n" if !defined $name;
    my $config = eval{ Config::Any->load_files({
        files => [$name],
        use_ext => 1,
        flatten_to_hash => 1,
    })->{$name} };
    die "Failed to load config [$name]\n" if $@ or !defined $config;
    return $config;
}

# interpolate the config vars
sub format_args {
    my $stub = shift;
    my $pre  = shift || ''; # maybe init_
    my $rv = sprintf $stub->{"${pre}format"},
        @{$stub->{"${pre}args"} || []};
    return split /\s+/, $rv;
}

# check if we have a program installed, and locate it
# borrowed from IPC::Cmd
sub can_run {
    my $command = shift;

    use Config;
    require File::Spec;
    require ExtUtils::MakeMaker;

    if( File::Spec->file_name_is_absolute($command) ) {
        return MM->maybe_command($command);
    }
    else {
        for my $dir (
            (split /\Q$Config{path_sep}\E/, $ENV{PATH}),
            File::Spec->curdir
        ) {           
            my $abs = File::Spec->catfile($dir, $command);
            return $abs if $abs = MM->maybe_command($abs);
        }
    }
}

# recursively merge two hashes together with right-hand precedence
# borrowed from Catalyst::Utils
sub merge_hashes {
    my ( $lefthash, $righthash ) = @_;
    return $lefthash unless defined $righthash;

    my %merged = %$lefthash;
    for my $key ( keys %$righthash ) {
        my $right_ref = ( ref $righthash->{ $key } || '' ) eq 'HASH';
        my $left_ref  = ( ( exists $lefthash->{ $key } && ref $lefthash->{ $key } ) || '' ) eq 'HASH';
        if( $right_ref and $left_ref ) {
            $merged{ $key } = merge_hashes(
                $lefthash->{ $key }, $righthash->{ $key }
            );
        }
        else {
            $merged{ $key } = $righthash->{ $key };
        }
    }

    return \%merged;
}

__END__

=head1 AUTHOR

Oliver Gorwits C<< <oliver@cpan.org> >>

=head1 COPYRIGHT & LICENSE

Copyright (c) The University of Oxford 2009.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut



lib/Net/Netfilter/NetFlow/Process.pm
package Net::Netfilter::NetFlow::Process;

use strict;
use warnings FATAL => 'all';

use base 'Exporter';
our @EXPORT = qw(
    conntrack_init
    ct2ft
    ptee
);

use POSIX; # core
use Time::HiRes 'gettimeofday'; # core
use IPC::Run 'run';
use Log::Dispatch::Config;
use Log::Dispatch::Configurator::Any;
use Net::Netfilter::NetFlow::Utils;
use Net::Netfilter::NetFlow::ConntrackFormat;

# poke conntrack kernel hooks into waking up (bug?)
sub conntrack_init {
    my $config = shift;
    my $conntrack = can_run($config->{conntrack}->{progname})
        or die "Failed to find a local copy of conntrack in the path\n";

    run [$conntrack, format_args($config->{conntrack}, 'init_')],
        '>', '/dev/null', '2>&1';
}

# convert the conntrack output to flow-tools CSV input format
sub ct2ft {
    my $config = shift;
    my $got_alrm = 0;
    my $tracker = {};

    # respond to SIGALRM (thanks go to perlipc man page)
    my $alrm_handler = sub { ++$got_alrm };
    # POSIX unmasks the sigprocmask properly
    my $action = POSIX::SigAction->new($alrm_handler);
    POSIX::sigaction(&POSIX::SIGALRM, $action);

    my $ttl = $config->{ct2ft}->{ttl} || 60 * 60 * 24 * 7; # seven days;
    alarm $ttl;

    # XXX alarm will not fire until we have input to process
    while (<>) {

        # pruge tracked connections older than TTL seconds
        if ($got_alrm) {
            alarm 0;
            foreach my $p (keys %{$tracker}) {
                foreach my $k (keys %{$tracker->{$p}}) {
                    delete $tracker->{$p}->{$k}
                        if $tracker->{$p}->{$k} < ($^T - $ttl);
                }
            }
            $got_alrm = 0;
            alarm $ttl;
        }

        chomp;
        s/[^\s\d.A-Z]//g;
        next if m/^\s+$/;
        my $line = $_; 
        my @fields = split /\s+/, $line;
        next unless scalar @fields > 12; 

        next unless $fields[1] =~ m/^(NEW|DESTROY)$/;
        my $mode = $1; 
        next unless $fields[2] =~ m/^(1|6|17)$/;
        my $proto = $1; 

        next if $proto == 1 and 
            (($fields[5] ne '8') and ($fields[6] ne '8')); # only interested in ECHO

        if ($mode eq 'NEW') {
            my $key = join ',', @fields[ @{$ct_new_key{$proto}} ];
            $tracker->{$proto}->{$key} = $fields[0];
            next;
        }   

        my $key = join ',', @fields[ @{$ct_destroy_key{$proto}} ];
        next unless exists $tracker->{$proto}->{$key};

        my ($start_secs, $start_micsecs) = split /\./, $tracker->{$proto}->{$key};
        my ($end_secs,   $end_micsecs)   = split /\./, $fields[0];

        # secs and nanosecs (^9) since 1970
        my ($unix_secs, $micsecs) = gettimeofday;
        my $unix_nsecs = $micsecs * 1_000;

        # millisecs (^3) since "boot"
        my $sysuptime = (($unix_secs - $^T) * 1_000) + int ($micsecs / 1_000);

        # flow start/end in millisecs since "boot"
        my $first = (($start_secs - $^T) * 1_000) + int ($start_micsecs / 1_000);
        my $last  = (($end_secs   - $^T) * 1_000) + int ($end_micsecs   / 1_000);

        for my $dir (qw( private_src public_src dst )) {
            my ($dpkts, $doctets, $srcaddr, $dstaddr, $srcport, $dstport)
                = @fields[ @{$ct_mask_fields{$proto}{$dir}} ];

            print join ',',
                $unix_secs,
                $unix_nsecs,
                $sysuptime,
                $config->{flow_send}->{args}->[0] || '127.0.0.1',
                $dpkts,
                $doctets,
                $first,
                $last,
                $srcaddr,
                $dstaddr,
                '0.0.0.0', # NEXTHOP
                0, # INPUT (SNMP idx)
                0, # OUTPUT (SNMP idx)
                $srcport || 0, # might be ICMP
                $dstport || 0, # might be ICMP
                $proto,
                0, # TOS
                0; # TCP_FLAGS
            print "\n";
        }   
    } # while (<>)
}

# set up output tee to local syslog, and next process in pipe
sub ptee {
    my $config = shift;
    Log::Dispatch::Config->configure_and_watch(
        Log::Dispatch::Configurator::Any->new($config->{ptee}->{conf}) );
    my $dispatcher = Log::Dispatch::Config->instance;

    while (<>) {
        $dispatcher->notice($_);
    }
}

__END__

=head1 AUTHOR

Oliver Gorwits C<< <oliver@cpan.org> >>

=head1 COPYRIGHT & LICENSE

Copyright (c) The University of Oxford 2009.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut



lib/Net/Netfilter/NetFlow/ConntrackFormat.pm
package Net::Netfilter::NetFlow::ConntrackFormat;

use strict;
use warnings FATAL => 'all';

use base 'Exporter';
our @EXPORT = qw(
    %ct_new_key
    %ct_destroy_key
    %ct_mask_fields
);

# TODO document and make user configurable

# 1:icmp - src,src,dst,id
# 6:tcp  - src,sport,src,sport,dst,dport
# 17:udp - src,sport,src,sport,dst,dport
# first src is private, second is public (post SNAT)

our %ct_new_key = (
    1  => [4,11,5,8],
    6  => [5,7,11,13,6,8],
    17 => [4,6,10,12,5,7],
);

our %ct_destroy_key = (
    1  => [3,11,4,7],
    6  => [3,5,10,12,4,6],
    17 => [3,5,10,12,4,6],
);

# dpkts, doctets, srcaddr, dstaddr, srcport, dstport
our %ct_mask_fields = (
    1 => {
        # field 17 does not exist
        private_src => [8,9,3,4,17,17],
        public_src  => [8,9,11,10,17,17],
        dst => [15,16,10,11,17,17],
    },
    6 => {
        private_src => [7,8,3,4,5,6],
        public_src  => [7,8,10,9,12,11],
        dst => [13,14,9,10,11,12],
    },
    17 => {
        private_src => [7,8,3,4,5,6],
        public_src  => [7,8,10,9,12,11],
        dst => [13,14,9,10,11,12],
    },
);

__END__

=head1 AUTHOR

Oliver Gorwits C<< <oliver@cpan.org> >>

=head1 COPYRIGHT & LICENSE

Copyright (c) The University of Oxford 2009.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut



Makefile.PL
use inc::Module::Install 0.67;
perl_version '5.008001';
  
name        'Net-Netfilter-NetFlow';
all_from    'script/nfflowd';

requires    'Config'              => 0;
requires    'POSIX'               => 0;
requires    'Exporter'            => 0;
requires    'Time::HiRes'         => 0;
requires    'Getopt::Long'        => 0;
requires    'File::Spec'          => 0;
requires    'ExtUtils::MakeMaker' => 0;

requires    'Log::Dispatch::Configurator::Any' => '1.0005';
requires    'Daemon::Generic' => 0;
requires    'File::Slurp'     => '9999.06';
requires    'IPC::Run'        => 0;
requires    'Config::Any'     => '0.15';
requires    'Config::General' => 0;

test_requires  'Test::More' => 0;
test_requires  'Test::Pod'  => 0;
test_requires  'Test::Pod::Coverage' => 0;

install_script  'script/nfflowd';

WriteAll;

